samples/csharp/getting-started/BinaryClassification_SpamDetection/SpamDetectionConsoleApp/Program.cs

```diff
@@ -36,41 +35,36 @@ static void Main(string[] args)
             MLContext mlContext = new MLContext();
 
             // Specify the schema for spam data and read it into DataView.
-            var data = mlContext.Data.ReadFromTextFile<SpamInput>(path: TrainDataPath, hasHeader: true, separatorChar: '\t');
+            var data = mlContext.Data.LoadFromTextFile<SpamInput>(path: TrainDataPath, hasHeader: true, separatorChar: '\t');
 
             // Create the estimator which converts the text label to boolean, featurizes the text, and adds a linear trainer.
-            var dataProcessPipeLine = mlContext.Transforms.CustomMapping<MyInput, MyOutput>(mapAction: MyLambda.MyAction, contractName: "MyLambda")
-                .Append(mlContext.Transforms.Text.FeaturizeText(outputColumnName: DefaultColumnNames.Features, inputColumnName: nameof(SpamInput.Message)));
-            
+            var dataProcessPipeLine = mlContext.Transforms.CustomMapping<InputRow, OutputRow>(CustomMappings.IncomeMapping, nameof(CustomMappings.IncomeMapping))
+                .Append(mlContext.Transforms.Text.FeaturizeText(outputColumnName: DefaultColumnNames.Features, inputColumnName: nameof(SpamInput.Message)))
+                                // The dataset we have is skewed, as there are many more non-spam messages than spam messages.
+                                // While our model is relatively good at detecting the difference, this skewness leads it to always
+                                // say the message is not spam. We deal with this by lowering the threshold of the predictor. In reality,
+                                // it is useful to look at the precision-recall curve to identify the best possible threshold.
+                .Append(mlContext.Transforms.Normalize(new NormalizingEstimator.LogMeanVarColumnOptions("LogMeanVarNormalized", DefaultColumnNames.Features, useCdf: true)))
```

Comment:
@kdcllc  Could you  please explain what normalize function does here and explain about the parameters?

Reply:
Though it can help in certain cases, you'll likely not want to renormalize the text column. The output of the `FeaturizeText()` runs though a LpNorm with defaults of L2 normalization. 

The row-wise normalization from the LpNorm and the column-wise LogMeanVar normalization can interact poorly leading to worse results on many datasets. I'd recommend not adding it.

Reply:
I might recommend a pipeline of:
```C#
// Data process configuration with pipeline data transformations 
var dataProcessPipeline = mlContext.Transforms.Conversion.MapValueToKey("Label", "Label")
                          .Append(mlContext.Transforms.Text.FeaturizeText("FeaturesText", new Microsoft.ML.Transforms.Text.TextFeaturizingEstimator.Options
                          {
                              WordFeatureExtractor = new Microsoft.ML.Transforms.Text.WordBagEstimator.Options { NgramLength = 2, UseAllLengths = true },
                              CharFeatureExtractor = new Microsoft.ML.Transforms.Text.WordBagEstimator.Options { NgramLength = 3, UseAllLengths = false },
                          }, "text"))
                          .Append(mlContext.Transforms.CopyColumns("Features", "FeaturesText"))
                          .Append(mlContext.Transforms.Normalize("Features", "Features"))
                          .AppendCacheCheckpoint(mlContext);

// Set the training algorithm 
var trainer = mlContext.MulticlassClassification.Trainers.OneVersusAll(mlContext.BinaryClassification.Trainers.AveragedPerceptron(labelColumnName: "Label", numberOfIterations: 10, featureColumnName: "Features"), labelColumnName: "Label")
                          .Append(mlContext.Transforms.Conversion.MapKeyToValue("PredictedLabel", "PredictedLabel"));
var trainingPipeline = dataProcessPipeline.Append(trainer);

return trainingPipeline;
```

Scores from this pipeline:
```md
*************************************************************************************************************
*       Metrics for Multi-class Classification model      
*------------------------------------------------------------------------------------------------------------
*       Average MicroAccuracy:    0.9899  - Standard deviation: (0.0050)  - Confidence Interval 95%: (0.0033)
*       Average MacroAccuracy:    0.9653  - Standard deviation: (0.0154)  - Confidence Interval 95%: (0.0101)
*       Average LogLoss:          0.0391  - Standard deviation: (0.0144)  - Confidence Interval 95%: (0.0094)
*       Average LogLossReduction: 0.9012  - Standard deviation: (0.0328)  - Confidence Interval 95%: (0.0214)
*************************************************************************************************************
```

To more easily handle the text labels, I'm running as multi-class. 

The label skew doesn't seem to have a large effect. The non-spam class is 87% of the dataset, and at around 99% accuracy, the model is well above just guessing the largest class.

Reply:
> Though it can help in certain cases, you'll likely not want to renormalize the text column. The output of the `FeaturizeText()` runs though a LpNorm with defaults of L2 normalization.
> 
> The row-wise normalization from the LpNorm and the column-wise LogMeanVar normalization can interact poorly leading to worse results on many datasets. I'd recommend not adding it.

Since in version 0.11.0 there is not way to adjust threshold. I was attempting to `renormalize` some of the data in the dataset. I am no Data Scientist. This is a good explanation of what I thought it might cause with larger dataset. Thank you.